File: LeetCode/25. K 个一组翻转链表.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
  ListNode *reverseKGroup(ListNode *head, int k)
  {
    ListNode *hair = new ListNode(0);
    hair->next = head;
    ListNode *pre = hair;
    while (head)
    {
      ListNode *tail = pre;
      for (int i = 0; i < k; i++)
      {
        tail = tail->next;
        if (!tail)
          return hair->next;
      }
      ListNode *nxt = tail->next;
      tie(head, tail) = reverseLink(head, tail);
      pre->next = head;
      tail->next = nxt;
      pre = tail;
      head = nxt;
    }
    return hair->next;
  }

private:
  pair<ListNode *, ListNode *> reverseLink(ListNode *head, ListNode *tail)
  {
    ListNode *pre = tail->next;
    ListNode *p = head;
    while (pre != tail)
    {
      ListNode *nxt = p->next;
      p->next = pre;
      pre = p;
      p = nxt;
    }
    return {tail, head};
  }
};

File: LeetCode/25. K 个一组翻转链表.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function (head, k) {
  let hair = new ListNode(0);
  hair.next = head;
  let pre = hair;
  while (head) {
    let tail = pre;
    for (let i = 0; i < k; i++) {
      tail = tail.next;
      if (!tail) return hair.next;
    }
    let nex = tail.next;
    [head, tail] = myReverse(head, tail);
    pre.next = head;
    tail.next = nex;
    pre = tail;
    head = tail.next;
  }
  return hair.next;
};

function myReverse(head, tail) {
  let pre = tail.next;
  let p = head;
  while (tail !== pre) {
    const nex = p.next;
    p.next = pre;
    pre = p;
    p = nex;
  }
  return [tail, head];
}
